### PR TITLE
90 base models

### PR DESCRIPTION
**Pull request by njuguna-n**
# Description

Add base models

#90 

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

**Comment by njuguna-n** [comment]
Changing the index type seems to have helped with the data_record model [issues](https://github.com/medic/cht-pipeline/pull/94#discussion_r1625774158) I had noted but still running a sync to confirm. Currently with about 4.5 million rows the average dbt run time is 2 minutes with a sync rate ~631 docs per second. I will provide a more comprehensive summary tomorrow once more docs have synced.

**Comment by njuguna-n** [comment]
Both DBT and postgres are running on EKS with the Postgres pod having 6GB in memory and 6 vCPUs resource limits and the DBT pod having 4GB of memory and 4 vCPUs.

The last successful run took ~3.5 hours and takes longer with more records with the data_record model still takes the longest to update. There are currently 32 million docs already synced.


```
20:35:16  1 of 4 START sql incremental model v1.stable_couchdb ........................... [RUN]
20:47:09  1 of 4 OK created sql incremental model v1.stable_couchdb ...................... [INSERT 0 2816944 in 713.54s]
20:47:09  2 of 4 START sql incremental model v1.contact .................................. [RUN]
20:49:54  2 of 4 OK created sql incremental model v1.contact ............................. [INSERT 0 154449 in 164.70s]
20:49:54  3 of 4 START sql incremental model v1.data_record .............................. [RUN]
00:02:09  3 of 4 OK created sql incremental model v1.data_record ......................... [INSERT 0 1189909 in 11534.79s]
00:02:09  4 of 4 START sql incremental model v1.person ................................... [RUN]
00:02:29  4 of 4 OK created sql incremental model v1.person .............................. [INSERT 0 123924 in 20.35s]
00:02:29
00:02:29  Finished running 4 incremental models in 3 hours 27 minutes and 13.54 seconds (12433.54s).
```

**Comment by witash** [comment]
@njuguna-n @andrablaj not going to mark this ready for review until tests are updated, but the latest version is ready, I have included a bunch of design notes for your review

Also @garethbowen if you are interested
### Naming Conventions
 - ids
	 - the PK for all incremental tables is the same as the source document `_id`.
	 - It is named `uuid` instead of id or `_id` to differentiate it from the other _ids which are not translated to primary keys (e.g. `patient_id` and `place_id`). This is arbitrary but somehow or other we need to distinguish between these two types of identifiers; if we just used `id`, we'd have `patient.id` and `patient.patient_id`which would also be arbitrary. IDK, I'm open to other suggestions. 
	 -  FKs use the convention `referencedtable_uuid` (instead of `referencedtable_id`) for the same reason; to distinguish between using the PK of another table or using the other id which is not a PK (see `data_record` which has `contact_uuid`, a FK to `contact` and also `patient_id` and `place_id` which are **not** FKs
- pluralization
	- table names are singular, it doesn't matter but we need to pick one or the other.

### Documents -> Tables
For contacts, one document in couchdb can translate to rows in several different tables, e.g. a patient document is a contact, a person, and a patient.  This allows queries for contacts to be simpler and more straightforward than if we tried to force one document to be one table (i.e. we want to be able to query patients by `SELECT * FROM patient` and get fields related to patients, instead of having of having all fields in a single contact table), but means we need to model parent-child relationships. 

 - each table has **only** the fields which are relevant to its own type. child tables filter the parent table by the conditions which identify the child type, and extract fields for that child type from the root table. (e.g. `person` filters `contact` where `type = 'person'` and extracts `date_of_birth` and `sex` from the couchdb doc.) 
 - This means child tables have to join back to the root table, but avoids having to copy the entire `doc` column into each model or have columns that are only relevant to child type in parent tables (e.g. clinics with birth dates)
 - `data_record` copies `fields` from the couchdb document. Form models are expected to use this rather than joining back to the root table; this is an exception to the above, and not strictly normalized because the same data is stored both in the root table and `data_record`. Most columns in form models will be extracted from `fields` so I thought it was worth it for clarity and so form views usually won't have to join back to the root table.

### PKs and parent FKS
 The parent child relationships where one document is multiple tables means that the ids and FKs look different than what you might expect if designing a relational database from scratch. 
 - each table has a PK `uuid` which is **both** the primary key and a FK to the parent table, and eventually, the `_id` of the source document, e.g. `person.uuid` is the same as `contact.uuid`. This means getting the contact details for a `patient` means joining on `patient.uuid = contact.uuid` rather than an explitly named `patient.contact_uuid`.
 - another option is to separate the PK and FK into different fields, but they would always have the same value so wouldn't be normalized.

### Contact Hierarchy Relationships
 There are relationships between different types of contacts, both globally, and specific to the configurable contact hierarchy, that these models do not attempt to model. A person's parent contact should be a place, a place's parent contact should be another place, or null if and only if the root, etc. but with these models a contacts parent is only another contact. 
 - This means that there is a no chw table. Is there any condition that can be used to reliably identify chws from other 'persons'? the brac sample used `where type = 'person'` and parent type = 'health_center', but that doesn't seem extensible.
 - The brac models renamed some of the contact fields, especially `chp.area_uuid` and `chp.region_uuid`, and then used these renamed fields very often. This seems like it would be useful for other partners as well, but I don't think including names for contact types in the base models is possible without denormalization or making assumptions about type relationships between contacts. So the recommendation is to add additional application specific models (views) that rename the parent contact fields based on the configurable hierarchy.

### Root table
This version does not have a dbt model for `couchdb`; `data_record` `contact`, `user` and the other tables (to be added) refer directly to the table that couch2pg is writing to, which is what I'm calling the 'root table'. In large databases, copying this table to a new dbt model basically doubles the storage space needed, (which can be significant), adds a lot of data copying to the dbt run, and isn't really necessary if the table couch2pg is writing to has all the necessary columns and indexes, but does mean that cht-pipeline relies on cht-sync to create this table with the correct indexes and columns

### Indexes
These models use btrees (the default) for any field other than a PK, where a hash index is appropriate. Other index types (brin, hash) may look useful on paper, but caused crippling performance issues in testing; the added benefit is probably not enough to outweigh the cost of further debugging to figure out why.

**Comment by witash** [comment]
a note about performance when reading from the root table (instead of having a `couchdb` initial model). It does mean that `type` is not indexed (although could always pull that out and index in cht-sync) which can lead to some strange results; it means the incremental queries always use the timestamp index; if a bunch of records were added to the root table with different types, the query planner can't see that and has to filter a bunch of records that are not relevant (e.g. if users haven't changed in a long time, but many new `data_records` are added, the query for users has to filter through all those new `data_records` without an index.)

I don't think its a problem, but if we notice some slowness (a few seconds vs. milliseconds) after all the data has been loaded to the couchdb table, especially for users, this might be why

**Comment by garethbowen** [comment]
@witash Thank you for the great writeup! The majority of that makes sense.

One thing I'm curious about is the performance impacts of using a uuid for the primary key as opposed to an auto incrementing int. This could impact both write performance (updating the index) and queries/joins. Did you investigate this option?

**Comment by witash** [comment]
> One thing I'm curious about is the performance impacts of using a uuid for the primary key as opposed to an auto incrementing int. This could impact both write performance (updating the index) and queries/joins. Did you investigate this option?

The motivation for using the uuid as the primary key as opposed to an auto incrementing int was to use the document _id from couchdb, and so not add another id column. I didn't check the performance specifically, from [googling around](https://www.cybertec-postgresql.com/en/int4-vs-int8-vs-uuid-vs-numeric-performance-on-bigger-joins/) I wouldn't expect it to make much difference, but we can do a quick test.

Speaking of testing, the brac results are available on the cht-sync-dev namespace @njogz

**Comment by njuguna-n** [comment]
@witash with some basic tests added are you now comfortable to merge this?

**Comment by witash** [comment]
> @witash with some basic tests added are you now comfortable to merge this?

yes I added some more tests that compare the root table to the dbt tables, it's ready to merge now

File: tests/sqltest/contact.sql [hunk 1]
```diff
@@ -0,0 +1,14 @@
+SELECT
+FROM v1.couchdb couchdb
```
Comment (njuguna-n) [inline tests/sqltest/contact.sql:2]: Can we use the values set in the environment variables for the schema and the root table here and in the other tests?

**Comment by njuguna-n** [comment]
@witash seeing as I opened the PR I am unable to approve it but I'm happy with the changes. I have set you as a reviewer.

**Review by witash — APPROVED** [review]

**Pull request merged**